### PR TITLE
Perftune - modified caching of flow xml descriptors, and added a faster cache lookup 

### PR DESCRIPTION
**Pull request by damonfeldman**
Used the faster cache lookup (which does not require the caller to create a function() / closure) for the flow lookup and also for debug and trace flag lookup. 

Overall, this contributes a minor performance improvement of a few seconds on my laptop.

Also added an XML/XQuery example to the Performance Test example.

### CODE REVIEW

**Comment by aebadirad** [comment]
There's a known bug in ml-gradle 3.2.0 that is fixed in 3.2.1, we may want to update this branch and bring it in, or pull this to another branch, perform the ml-gradle upgrade there, and then pull to develop. @paxtonhare @dmcassel @damonfeldman

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/debug-lib.xqy [hunk 1]
```diff
@@ -47,6 +47,28 @@ declare function debug:enable($enabled as xs:boolean)
...
+              fn:doc("/com.marklogic.hub/settings/__debug_enabled__.xml"),
+              cts:element-value-query(xs:QName("debug:is-debugging-enabled"), "1", ("exact")),
+              ("unfiltered", "score-zero", "unchecked", "unfaceted")
+            )
+          )
+        ',(), map:new(map:entry("database", xdmp:modules-database()))),
+        ()
+      )
+
+  (:
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/debug-lib.xqy]: Can you remove the commented out code? No need to keep it around.

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/hub-utils-lib.xqy [hunk 2]
```diff
@@ -86,21 +86,44 @@ declare function hul:invalidate-field-cache($name)
...
-    if (fn:exists($existing)) then
-      if (fn:exists($refresh-date) and ($refresh-date < fn:current-dateTime())) then
-        hul:set-field-cache($name, $func(), $duration)
-      else
+    if ( (exists($refresh-date) and ($refresh-date < fn:current-dateTime() + xs:dayTimeDuration("PT0."||xdmp:random(9)||"S")))
+         or empty($existing)
+         or exists($duration) and empty($refresh-date) 
+        )
+      then 
+        ( xdmp:log("field cache miss xx : "||$name),
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/hub-utils-lib.xqy]: generally we've been avoiding xdmp:logs(). Do we need to log this out? or is this leftover debugging code?

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/trace-lib.xqy [hunk 3]
```diff
@@ -73,6 +73,29 @@ declare function trace:enable-tracing($enabled as xs:boolean)
...
+              cts:element-value-query(xs:QName("trace:is-tracing-enabled"), "1", ("exact")),
+              ("unfiltered", "score-zero", "unchecked", "unfaceted")
+            )
+          ) 
+        ',(), map:new(map:entry("database", xdmp:modules-database()))),
+        xs:dayTimeDuration("PT1M")
+      )
+
+
+(:
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/trace-lib.xqy]: similar to above.. please remove the old code in the comment.

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/mlcp-flow-transform.xqy [hunk 4]
```diff
@@ -78,7 +78,8 @@ declare function mlcpFlow:transform(
 declare function mlcpFlow:run-flow(
   $jobId, $flow, $uri, $content, $options)
 {
-  (: mlcp in runs in update mode :)
+  try{
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/mlcp-flow-transform.xqy]: the old code would not catch an exception and the error would bubble up to mlcp. Why the try/catch here?

File: marklogic-data-hub/build.gradle [hunk 5]
```diff
@@ -3,7 +3,7 @@ plugins {
     id 'java'
     id 'maven-publish'
     id 'com.jfrog.bintray' version '1.7.2'
-    id 'com.marklogic.ml-gradle' version '3.1-alpha4'
+    id 'com.marklogic.ml-gradle' version '3.2.0'
```
Comment (paxtonhare) [inline marklogic-data-hub/build.gradle]: To @aebadirad 's point.. we plan to update to the latest ml-gradle in the next sprint. Can you please set this back to 3.1-alpha4 for now?

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/mlcp-flow-transform.xqy (hunk 4 above)
Comment (damonfeldman) [reply marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/mlcp-flow-transform.xqy]: I think mlcp was showing a dummy line like "there was an error" but the stack/details were lost, so I added that while debugging an issue, and left it in as useful.

**Comment by damonfeldman** [comment]
I've pushed some changes to my perftune branch, on the theory that this will show automatically in this pull request. Please review. Tests run.

**Comment by aebadirad** [comment]
I've done the rebase and manual merge, and reopened the PR in #725

**Pull request closed without merging**